Add feedback to admin menu

JoeyStk opened the pull request:
### Short description
This PR is still a work in progress. This PR in it's entirety is supposed to create an api endpoint to allow the lunes app to send feedback. 

### Proposed changes
This is the first out of two commits. This commit is supposed to
- add a feedback item to the admin menu

But at even at this point I'm having issues. There is something wrong with my classes Feedback and FeedbackAdmin. Therefore the iteration of the register-function is failing. Can you maybe give me a hint how to fix this respectively what is still missing in order to add "feedback" to the admin menu. 

![grafik](https://user-images.githubusercontent.com/72705147/170713419-55706e35-0e88-429e-9924-b9047ebfc10c.png)

### Resolved issues
<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #211 
(for more information also take a look at #313 )

timoludwig commented on lunes_cms/cms/admin.py:
```diff
@@ -76,3 +78,4 @@ def get_app_list(self, request):
 admin.site.register(Document, DocumentAdmin)
 admin.site.register(GroupAPIKey, GroupAPIKeyAdmin)
 admin.site.unregister(APIKey)
+admin.site.register(FeedbackAdmin)
```
```suggestion
admin.site.register(Feedback, FeedbackAdmin)
```
You need to pass the model to the register function as well.
This requires the model to be imported in line 11 obviously:
```
from .models import Discipline, TrainingSet, Document, GroupAPIKey, Feedback
```

timoludwig commented on lunes_cms/cms/admins/feedback_admin.py:
```diff
@@ -0,0 +1,10 @@
+    Admin interface for feedback.
+    Inheriting from `admin.ModelAdmin`.
+    """
+
```
```suggestion

    model = Feedback

```
I guess you should define the model here as well.

timoludwig commented on lunes_cms/cms/admins/feedback_admin.py:
```diff
@@ -0,0 +1,10 @@
+from django.contrib import admin
+
```
```suggestion

from ..models import Feedback


```

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,25 @@
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+
```
```suggestion

MATTER_CHOICES = (
    ("GENERAL", _("general")),
    ("DISCIPLINE", _("discipline")),
    ("TRAINING_SET", _("training set")),
    ("DOCUMENT", _("document")),
)


```
I'd move the matter choices outside of the model definition.
Also, the second part of the tuple is usually the "readable" part, so it makes sense to translate them.

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,25 @@
+    Contains a title, a description and a reference to the matter of feedback
+    """
+
+    id = models.AutoField(primary_key=True)
```
```suggestion
```
The id field is not required, it is automatically set.

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,25 @@
+    id = models.AutoField(primary_key=True)
+    title = models.CharField(max_length=255)
+    description = models.TextField()
+    MATTER_CHOICES = (
+        ('general','GENERAL'), 
+        ('training_set','TRANING_SET'), 
+        ('document','DOCUMENT'), 
+        ('user_group','USER_GROUP')
+    )
```
```suggestion
```

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,25 @@
+        ('document','DOCUMENT'), 
+        ('user_group','USER_GROUP')
+    )
+    matter = models.ForeignKey(max_length=4, choices=MATTER_CHOICES, default='general')
```
```suggestion
    matter = models.CharField(
        max_length=12, choices=MATTER_CHOICES, default=MATTER_CHOICES[0][0]
    )
```
A `ForeignKey` is a link to another model. If you want to store constant strings in this field, you need a `CharField`.
Also, it makse sense not to hardcode the default value, but to reference the choice definition.

timoludwig commented on lunes_cms/locale/de/LC_MESSAGES/django.po:
```diff
@@ -35,6 +35,10 @@ msgstr "Vokabeln"
 
+#: cms/admin.py:36
+msgid "feedback"
+msgstr ""
```
```suggestion
msgstr "Feedback"
```
Even though we don't translate this string, our translation checking tool expects something here.

timoludwig reviewed (commented):
Thanks, this is going in the right direction. :+1:
Se my comments below for initial feedback.

JoeyStk commented:
Thank you a lot for your feedback! It helped me a lot.
I worked on the next two steps: 
I finished the commit to add feedback to the navigation (admin). In the end it worked fine on my machine. Please feel free to test if it's working as expected.
Then I started working on the API. However I'm still struggling to understand ViewSets. Neither the REST framework documentation nor tutorials helped me really understand them. Can you maybe give a few guidelines on what to focus when I'm working with them?
Lastly the black-test is still failing. The documentation in Circle CI doesn't help me a lot. It only tells me which files aren't well formatted. But still I'm not sure what to fix in order to pass the black-test. Can you maybe give me some hints on how to troubleshoot?
Thank you again for helping me out! I really appreciate it!

timoludwig commented on lunes_cms/cms/serializers.py:
```diff
@@ -170,6 +170,22 @@ class Meta:
+
+        model = Feedback
+        fields = (
+            ""
```
```suggestion
            "title",
            "description",
            "matter",
```

timoludwig commented on lunes_cms/cms/views/feedback_viewset.py:
```diff
@@ -0,0 +1,6 @@
+from rest_framework import viewsets
+from ..models import Feedback
+from ..serializers import FeedbackSerializer
+
+# class FeedbackViewSet(viewsets.ModelViewSet):
+    
```
```suggestion
from rest_framework import mixins, viewsets
from ..serializers import FeedbackSerializer

class CreateFeedbackViewSet(mixins.CreateModelMixin, viewsets.GenericViewSet):
    """
    A viewset that provides only the ``create`` action for feedback elements.
    """
    serializer_class = FeedbackSerializer

```

timoludwig reviewed (commented):
> Can you maybe give a few guidelines on what to focus when I'm working with them?

The viewsets provide an easy way to define multiple API endpoints at once, so the complete CRUD functionality only needs one ViewSet class. Since in our case, we only need the "create" functionality and do not care about read/update/delete, we can decide whether we want to use an old-school simple Django-view as endpoint, or whether we limit the viewset to only the create functionality. The latter is probably a bit easier, so I'd suggest we try this first:

```
from rest_framework import mixins, viewsets
from ..serializers import FeedbackSerializer

class CreateFeedbackViewSet(mixins.CreateModelMixin, viewsets.GenericViewSet):
    """
    A viewset that provides only the ``create`` action for feedback elements.
    """
    serializer_class = FeedbackSerializer
```

Similar to the "Custom ViewSet base classes" mentioned here: https://www.django-rest-framework.org/api-guide/viewsets/#custom-viewset-base-classes

Then, we should import this viewset in [lunes_cms/cms/views/__init__.py](https://github.com/digitalfabrik/lunes-cms/blob/develop/lunes_cms/cms/views/__init__.py):
```
from .feedback_viewset import CreateFeedbackViewSet
```


After that, we just need to register the route to this viewset in the [urls.py](https://github.com/digitalfabrik/lunes-cms/blob/develop/lunes_cms/cms/urls.py):
```
router.register(r"feedback", views.CreateFeedbackViewSet, "feedback")
```
Now we're done and we can try to create feedback elements via the API:
1. Via the web-interface:
   http://localhost:8080/api/feedback
2. Via the API:
    ```
    curl -X POST http://localhost:8080/api/feedback -H 'Content-Type: application/json' -d '{"title": "test", "description": "works well", "matter": "DISCIPLINE"}'
    ```


<hr>

> Can you maybe give me some hints on how to troubleshoot?

Good point, I think we are missing a dev tool to apply the black formatting. For the time being, you can do one of two things:

1. Enable the pre-commit hooks (this will cause black to run on every commit in the future):
    ```
    # Activate virtual environment
    source .venv/bin/activate
    # Install pre-commit hooks
    pre-commit install
    # Run pre-commit hooks without doing actual commit
    pre-commit run --all-files
    ```

3. Run black tool manually
    ```
    # Activate virtual environment
    source .venv/bin/activate
    # Apply black formatting
    black .
    ```

JoeyStk commented:
Thank you again for the detailed feedback! It once again helped me a lot! :)

I see still three points left, before I can show the new endpoint to the App-Team. 
1. I tried the idea to solve the matter via a GenericForeignKey, but it didn't work at all, as I imagined it. For one the import of GenericForeignKey doesn't seem to exist.

2. I tried to make the feedback only visible to the administrators until we come up with a solution on how to display the feedback in a way that makes sense. Is there another way to do this instead of using "only_staff"?

3. When I try to start the server, I'm running into the following error message. I'm quite sure this error appeared when I updated the migration file for feedback respectively when I rebased from development and added the migration file 003. I already tried to delete the migration files and the pycache. And although that helped to clean up a bit, it didn't fix the error. Do you have an idea what I could do to fix this? 
![grafik](https://user-images.githubusercontent.com/72705147/173085406-5c84a7a1-3de3-4a5b-b602-14f08f96c3d0.png)

timoludwig commented on lunes_cms/cms/admins/feedback_admin.py:
```diff
@@ -0,0 +1,16 @@
+    """
+
+    model = Feedback
+
```
```suggestion
    list_display = readonly_fields = [
        "comment",
        "content_object_link",
        "content_object_type",
        "created_date",
        "read_by",
    ]
    search_fields = ["comment"]
    list_filter = ["content_type", "read_by"]
    sortable_by = ["content_object_type", "created_date", "read_by"]

    def has_add_permission(self, request, obj=None):
        return False

    def has_change_permission(self, request, obj=None):
        return False

    class Media:
        css = {"all": ("css/feedback.css",)}
```
This would be my suggestion for the feedback admin config, but you can have a look yourself whether you want to change the displayed fields etc...
Also, I think we can resolve the "read-only" part by just returning `False` for the add/change permissions...
And I'd like to add a bit of custom css to this list, since we don't know how much content the feedback entries might contain (and thus we don't know the table row width).

timoludwig commented on lunes_cms/cms/admins/feedback_admin.py:
```diff
@@ -0,0 +1,16 @@
+
+    model = Feedback
+
+    def save_model(self, request, obj, form, change):
+        obj.save()
```
```suggestion
```
Since we're not saving anything, I guess we can delete this.

timoludwig commented on lunes_cms/cms/migrations/0004_feedback.py:
```diff
@@ -0,0 +1,46 @@
+
+    dependencies = [
+        ("cms", "0003_remove_documentimage_name"),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name="Feedback",
+            fields=[
+                (
+                    "id",
+                    models.AutoField(
+                        auto_created=True,
+                        primary_key=True,
+                        serialize=False,
+                        verbose_name="ID",
+                    ),
+                ),
+                ("title", models.CharField(max_length=255)),
+                ("description", models.TextField()),
+                (
+                    "matter",
+                    models.CharField(
+                        choices=[
+                            ("GENERAL", "general"),
+                            ("DISCIPLINE", "discipline"),
+                            ("TRAINING_SET", "training set"),
+                            ("DOCUMENT", "document"),
+                        ],
+                        default="GENERAL",
+                        max_length=12,
+                    ),
+                ),
+            ],
+            options={
+                "verbose_name": "feedback",
+                "verbose_name_plural": "feedback",
+            },
+        ),
+    ]
```
```suggestion
from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
        ("contenttypes", "0002_remove_content_type_name"),
        ("cms", "0003_remove_documentimage_name"),
    ]

    operations = [
        migrations.CreateModel(
            name="Feedback",
            fields=[
                (
                    "id",
                    models.AutoField(
                        auto_created=True,
                        primary_key=True,
                        serialize=False,
                        verbose_name="ID",
                    ),
                ),
                (
                    "object_id",
                    models.PositiveIntegerField(
                        help_text="The id of the object this feedback entry refers to.",
                        verbose_name="object id",
                    ),
                ),
                ("comment", models.TextField(verbose_name="comment")),
                (
                    "created_date",
                    models.DateTimeField(
                        auto_now_add=True,
                        help_text="The time and date when the feedback was submitted.",
                        verbose_name="submitted on",
                    ),
                ),
                (
                    "content_type",
                    models.ForeignKey(
                        help_text="The content type this feedback entry refers to.",
                        limit_choices_to=models.Q(
                            ("app_label", "cms"),
                            ("model__in", ["discipline", "trainingset", "document"]),
                        ),
                        on_delete=django.db.models.deletion.CASCADE,
                        to="contenttypes.contenttype",
                        verbose_name="content type",
                    ),
                ),
                (
                    "read_by",
                    models.ForeignKey(
                        blank=True,
                        help_text="The user who marked this feedback as read. If the feedback is unread, this field is empty.",
                        null=True,
                        on_delete=django.db.models.deletion.SET_NULL,
                        related_name="feedback",
                        to=settings.AUTH_USER_MODEL,
                        verbose_name="marked as read by",
                    ),
                ),
            ],
            options={
                "verbose_name": "feedback",
                "verbose_name_plural": "feedback entries",
            },
        ),
    ]
```
I'm not sure what went wrong with your migration, but when I apply my suggestions and re-generate the migration file, this is the result :see_no_evil:

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+# from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.utils.translation import ugettext_lazy as _
+from django.db import models
```
```suggestion
from html import escape

from django.conf import settings
from django.contrib.contenttypes.fields import GenericForeignKey
from django.contrib.contenttypes.models import ContentType
from django.db import models
from django.urls import reverse
from django.utils.translation import gettext_lazy as _
from django.utils.safestring import mark_safe
from django.utils.text import Truncator
```

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+
+class Feedback(models.Model):
+    """
+    Contains a title, a description and a reference to the matter of feedback
```
```suggestion
    Model to store feedback about disciplines, training sets and vocabulary words
```

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+    Contains a title, a description and a reference to the matter of feedback
+    """
+
+    title = models.CharField(max_length=255)
```
```suggestion
```
I don't think we need a title... probably, users will only send one comment as feedback...

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+    """
+
+    title = models.CharField(max_length=255)
+    description = models.TextField()
+    discipline = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    document = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    training_set = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    # options = models.GenericForeignKey("general", discipline, document, training_set)
+    # matter = models.CharField(max_length=12, choices=options, default=options[0])
```
```suggestion
    content_type = models.ForeignKey(
        ContentType,
        limit_choices_to=models.Q(
            app_label="cms", model__in=["discipline", "trainingset", "document"]
        ),
        on_delete=models.CASCADE,
        verbose_name=_("content type"),
        help_text=_("The content type this feedback entry refers to."),
    )
    object_id = models.PositiveIntegerField(
        verbose_name=_("object id"),
        help_text=_("The id of the object this feedback entry refers to."),
    )
    content_object = GenericForeignKey("content_type", "object_id")
    content_object.short_description = _("refers to")

    comment = models.TextField(
        verbose_name=_("comment"),
    )

    read_by = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="feedback",
        verbose_name=_("marked as read by"),
        help_text=_(
            "The user who marked this feedback as read. If the feedback is unread, this field is empty."
        ),
    )

    created_date = models.DateTimeField(
        auto_now_add=True,
        verbose_name=_("submitted on"),
        help_text=_("The time and date when the feedback was submitted."),
    )
```
I think you misunderstood how the `GenericForeignKey` works, you don't need one field per model you want to link, but you need one field to store the type of the object (`content_type`) and one field to store the id of the object (`object_id`). Those two fields combined yields the generic foreign key `content_object` which can then be used like a regular foreign key (you can set e.g. `feedback.content_object = Discipline.objects.first()` and Django will automatically set `feedback.content_type` to `discipline` and `feedback.object_id` to the id of the first discipline). So the beauty of this feature is that independently of how many different content types you want to reference, you only need the two database fields and the one generic foreign key which handles the conversion.

Apart from that, I copied + pasted a few other fields from our Integreat feedback model: `read_by` and `created_date`.

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+    training_set = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    # options = models.GenericForeignKey("general", discipline, document, training_set)
+    # matter = models.CharField(max_length=12, choices=options, default=options[0])
+
```
```suggestion

    def content_object_link(self):
        """
        Include a link to the edit form of the content object in list display

        :return: Link to the edit form of the content object
        :rtype: str
        """
        # Get link to the admin form if the content object
        admin_edit_link = reverse(
            f"admin:cms_{self.content_type.model}_change", args=[self.object_id]
        )
        # Escape label because string is marked as safe afterwards
        escaped_label = escape(str(self.content_object))
        # Mark string as safe to allow html link tag
        return mark_safe(f"<a href={admin_edit_link}>{escaped_label}</a>")

    content_object_link.short_description = _("refers to")

    def content_object_type(self):
        """
        Better readable representation of the content type

        :return: Verbose name of the content object's model
        :rtype: str
        """
        return self.content_object._meta.verbose_name

    content_object_type.short_description = _("content type")
    content_object_type.admin_order_field = "content_type"

```
We could add two read only fields to improve the appearance of the generic foreign key:
- `content_object_link` provides a link to the edit form of the referenced object, so it can directly be improved from the feedback list/entry
- `content_object_type` provides a better representation of the object type

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+    # matter = models.CharField(max_length=12, choices=options, default=options[0])
+
+    def __str__(self):
+        """String representation of Feedback instance
+
+        :return: title of feedback instance
```
```suggestion
        """
        String representation of feedback instance

        :return: Truncated feedback comment
```

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+        :return: title of feedback instance
+        :rtype: str
+        """
+        return self.title
```
```suggestion
        return Truncator(self.comment).chars(80)
```
Instead of the title, we could just truncate the comment to a given number of characters

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+        return self.title
+
+    class Meta:
+        """
+        Define user readable name of Feedback
+        """
+
```
```suggestion
        #: The verbose name of the model
```
I'd say the inner Meta class doesn't need a class based docstring, but we can document it's attributes

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,34 @@
+        """
+
+        verbose_name = _("feedback")
+        verbose_name_plural = _("feedback")
```
```suggestion
        #: The plural verbose name of the model
        verbose_name_plural = _("feedback entries")
```
I'd go with e.g. "4 feedback entries" for the plural of feedback, otherwise "4 feedback" looks a bit strange.

timoludwig commented on lunes_cms/cms/serializers.py:
```diff
@@ -171,6 +178,25 @@ class Meta:
+
+        model = Feedback
+        fields = (
+            "title",
+            "description",
+            "matter",
```
```suggestion
            "comment",
            "content_type",
            "object_id",
```

timoludwig commented on lunes_cms/cms/views/feedback_viewset.py:
```diff
@@ -0,0 +1,11 @@
+    """
+
+    serializer_class = FeedbackSerializer
+    http_method_names = ["post"]
```
```suggestion
```
This is already implicitly solved by only inheriting the `CreateModelMixin` and the `GenericViewSet`. All methods that don't create are automatically forbidden.

timoludwig commented on lunes_cms/locale/de/LC_MESSAGES/django.po:
```diff
@@ -353,6 +357,9 @@ msgstr ""
 msgid "Summary"
 msgstr "Zusammenfassung"
 
+#~ msgid "general"
+#~ msgstr "Generelles Feedback"
```
```suggestion
#: cms/models/feedback.py:24 cms/models/feedback.py:83
msgid "content type"
msgstr "Inhaltstyp"

#: cms/models/feedback.py:25
msgid "The content type this feedback entry refers to."
msgstr "Der Inhaltstyp, auf den sich dieser Feedback-Eintrag bezieht."

#: cms/models/feedback.py:28
msgid "object id"
msgstr "Objekt-ID"

#: cms/models/feedback.py:29
msgid "The id of the object this feedback entry refers to."
msgstr "Die ID des Objekts, auf das sich dieser Feedback-Eintrag bezieht."

#: cms/models/feedback.py:32 cms/models/feedback.py:72
msgid "refers to"
msgstr "bezieht sich auf"

#: cms/models/feedback.py:35
msgid "comment"
msgstr "Kommentar"

#: cms/models/feedback.py:44
msgid "marked as read by"
msgstr "als gelesen markiert von"

#: cms/models/feedback.py:46
msgid ""
"The user who marked this feedback as read. If the feedback is unread, this "
"field is empty."
msgstr ""
"Der Benutzer, der dieses Feedback als gelesen markiert hat. Wenn das "
"Feedback ungelesen ist, ist dieses Feld leer."

#: cms/models/feedback.py:52
msgid "submitted on"
msgstr "übermittelt am"

#: cms/models/feedback.py:53
msgid "The time and date when the feedback was submitted."
msgstr "Uhrzeit und Datum, an dem das Feedback eingereicht wurde."

#: cms/models/feedback.py:99
msgid "feedback entries"
msgstr "Feedback-Einträge"
```

timoludwig reviewed (commented):
Thanks a lot, this is becoming better and better :rocket: :muscle: 

> 1. I tried the idea to solve the matter via a GenericForeignKey, but it didn't work at all, as I imagined it. For one the import of GenericForeignKey doesn't seem to exist.

The model fields were defined a bit incorrectly, see my comment below for details.

> 2. Is there another way to do this instead of using "only_staff"?

In fact, restricting the access to staff members does not help here because all members that have access to the admin area are staff members. However, I think this is automatically solved as long as we don't grant users the permission to view feedback explicitly.

> 3. Do you have an idea what I could do to fix this?

I did not have this problem, so I don't think it is related to the code base or the migration files.
Probably you need to delete your local database with `rm lunes_cms/db.sqlite3` and start the server again.


Apart from that, I have two additional ideas:
1. You could add the line `"cms.Feedback": "fas fa-comment",` to the [jazzmin icon settings](https://github.com/digitalfabrik/lunes-cms/blob/eed6b36fe5ca7f38197c0021d42a4567b7b5b2f6/lunes_cms/core/settings.py#L418) to customize the icon for the feedback menu entry
2. We could make sure very long feedback entries don't mess up the table by creating the following CSS file:
<details>
  <summary>lunes_cms/cms/static/css/feedback.css</summary>

  ```
/* Additional CSS for styling the feedback table */

table {
    width: 100%;
    table-layout: fixed;
    overflow-wrap: break-word;
    min-width: 1000px;
}

/* Checkbox column */
thead th:nth-child(1) {
    width: 40px;
}

/* Comment column */
thead th:nth-child(2) {
    width: auto;
}

/* Related object column */
thead th:nth-child(3) {
    width: 250px;
}

/* Related object type column */
thead th:nth-child(4) {
    width: 140px;
}

/* Submission date column */
thead th:nth-child(5) {
    width: 185px;
}

/* Read by column */
thead th:nth-child(6) {
    width: 230px;
}

/* Truncate long comments and content object links */
.field-comment a, .field-content_object_link a {
    display: block;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
 }

  ```

</details>

JoeyStk commented:
Thank you once again for your feedback! :)
I was able to fix the issue with my run-command. I still get one last issue. When I open the feedback-tab I get "no such column: cms_feedback.content_type_id". I'm not sure how this issue comes to be. I couldn't find any instance of "content_type_id". Thus, I'm now expecting it to be an error on my machine. Can you please check, if everything is working as expected on your machine? 
Thank you :)

timoludwig commented on lunes_cms/cms/models/feedback.py:
```diff
@@ -0,0 +1,99 @@
+        return mark_safe(f"<a href={admin_edit_link}>{escaped_label}</a>")
+
+    content_object_link.short_description = _("refers to")
+
+    def content_object_type(self):
+        """
+        Better readable representation of the content type
+
+        :return: Verbose name of the content object's model
+        :rtype: str
+        """
+        return self.content_object._meta.verbose_name
+
+    content_object_type.short_description = _("content type")
+    content_object_type.admin_order_field = "content_type"
```
```suggestion
```
I think I noticed a better way to do this - probably, we should just overwrite the `__str__()`-method of the content type model by creating a new file `lunes_cms/cms/models/content_type.py` with the following content:
```
from django.contrib.contenttypes.models import ContentType

# Better representation of content type
ContentType.add_to_class("__str__", lambda c: c.name)
```
and then 
```
from . import group, content_type
```
here:
https://github.com/digitalfabrik/lunes-cms/blob/eed6b36fe5ca7f38197c0021d42a4567b7b5b2f6/lunes_cms/cms/models/__init__.py#L7

timoludwig commented on lunes_cms/cms/admins/feedback_admin.py:
```diff
@@ -0,0 +1,32 @@
+    list_display = readonly_fields = [
+        "comment",
+        "content_object_link",
+        "content_object_type",
```
```suggestion
        "content_type",
```

timoludwig commented on lunes_cms/cms/admins/feedback_admin.py:
```diff
@@ -0,0 +1,32 @@
+    ]
+    search_fields = ["comment"]
+    list_filter = ["content_type", "read_by"]
+    sortable_by = ["content_object_type", "created_date", "read_by"]
```
```suggestion
    sortable_by = ["content_type", "created_date", "read_by"]
```

timoludwig commented on lunes_cms/cms/static/css/feedback.css:
```diff
@@ -0,0 +1,46 @@
+    min-width: 1000px;
+  }
+  
+  /* Checkbox column */
+  thead th:nth-child(1) {
+    width: 40px;
+  }
+  
+  /* Comment column */
+  thead th:nth-child(2) {
+    width: auto;
+  }
+  
+  /* Related object column */
+  thead th:nth-child(3) {
+    width: 250px;
+  }
+  
+  /* Related object type column */
+  thead th:nth-child(4) {
+    width: 140px;
+  }
+  
+  /* Submission date column */
+  thead th:nth-child(5) {
+    width: 185px;
+  }
+  
+  /* Read by column */
+  thead th:nth-child(6) {
+    width: 230px;
+  }
+  
+  /* Truncate long comments and content object links */
+  .field-comment a, .field-content_object_link a {
+    display: block;
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
+  }
```
```suggestion
/* Additional CSS for styling the feedback table */

table {
  width: 100%;
  table-layout: fixed;
  overflow-wrap: break-word;
  min-width: 1000px;
}

/* Checkbox column */
thead th:nth-child(1) {
  width: 40px;
}

/* Comment column */
thead th:nth-child(2) {
  width: auto;
}

/* Related object column */
thead th:nth-child(3) {
  width: 250px;
}

/* Related object type column */
thead th:nth-child(4) {
  width: 140px;
}

/* Submission date column */
thead th:nth-child(5) {
  width: 185px;
}

/* Read by column */
thead th:nth-child(6) {
  width: 230px;
}

/* Truncate long comments and content object links */
.field-comment a,
.field-content_object_link a {
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

```
Ok, I think the indentation of my previous suggestion was a bit weird :sweat_smile:

timoludwig commented on lunes_cms/cms/serializers.py:
```diff
@@ -171,6 +178,25 @@ class Meta:
+    Serializer for the Feedback module. Inherits from
+    'serializers.ModelSerializer'.
+    """
+
```
```suggestion

    content_type = serializers.SlugRelatedField(
        slug_field="model",
        queryset=ContentType.objects.filter(
            app_label="cms", model__in=["discipline", "trainingset", "document"]
        ),
        error_messages={
            "does_not_exist": _(
                "The content type must be either 'discipline', 'training set' or 'document'."
            ),
        },
    )

    def validate(self, attrs):
        try:
            attrs["content_type"].model_class().objects.get(pk=attrs["object_id"])
        except:
            raise serializers.ValidationError(
                {
                    "object_id": [
                        _("{} with id {} does not exist.").format(
                            attrs["content_type"].name, attrs["object_id"]
                        ),
                    ]
                }
            )
        return attrs

```
I noticed two further problems:
1. When the content type is referenced by id, the app somehow has to know which ids the content types have, e.g. 1 is discipline, 2 is document and 3 is training set... instead, we should serialize the content type as string, which can be achieved by using a `SlugRelatedField`.
2. At the moment it is possible to create feedback for non-existing object, which is not what we want I think. I would add a validator to make sure this does not happen.

timoludwig commented on lunes_cms/locale/de/LC_MESSAGES/django.po:
```diff
@@ -217,6 +221,54 @@ msgstr "Erstellt am"
+#: cms/models/feedback.py:99
+msgid "feedback entries"
+msgstr "Feedback-Einträge"
+
```
```suggestion

#: cms/serializers.py:196
msgid ""
"The content type must be either 'discipline', 'training set' or 'document'."
msgstr ""
"Der Inhaltstyp muss entweder 'discipline', 'trainingset' oder 'document' "
"sein."

#: cms/serializers.py:208
msgid "{} with id {} does not exist."
msgstr "{} mit der ID {} existiert nicht."

```

timoludwig reviewed (commented):
Thanks! :+1:

> When I open the feedback-tab I get "no such column: cms_feedback.content_type_id". I'm not sure how this issue comes to be.

For me, everything works, so seems to be a problem with your database again. In doubt, you can always delete the sqlite file again and restart the server, this should always apply the migrations on a fresh database.

Apart from that:
I think it would be cool if feedback entries would be cascade deleted if the related objects are deleted - since we don't need feedback about non-existing elements.
To do so, you have to add the following three lines to the models discipline, document and training_set:
```
from django.contrib.contenttypes.fields import GenericRelation
from .feedback import Feedback
```
and as last model field inside the class:
```
feedback = GenericRelation(Feedback)
```

Aaaand I think this change justifies an entry in the [changelog](https://github.com/digitalfabrik/lunes-cms/blob/develop/CHANGELOG.md) :slightly_smiling_face:

timoludwig reviewed (approved):
Awesome, I think that's it! :rocket: 
Let's merge this and deploy it to the test system so we can show it off to the app team.
